src/Microsoft.DotNet.XHarness.Common/Logging/ScanLog.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+#nullable enable
+namespace Microsoft.DotNet.XHarness.Common.Logging
+{
+    /// <summary>
+    /// Log that scans for a given tag and notifies when tag is found in the stream.
+    /// </summary>
+    public class ScanLog : Log
+    {
+        private readonly string _tag;
+        private readonly Action _tagFoundNotify;
+        private readonly char[] _buffer;
+        private int _startIndex;
+        private bool _hasBeenFilled = false;
+
+        public override bool Timestamp { get => false; set { } }
+
+        public ScanLog(string tag, Action tagFoundNotify)
+        {
+            _tag = tag;
+            _tagFoundNotify = tagFoundNotify;
+            _buffer = new char[_tag.Length];
+            _startIndex = -1;
+        }
+
+        protected override void WriteImpl(string value)
+        {
+            foreach (var c in value)
+            {
+                Add(c);
+
+                if (IsMatch())
+                {
+                    _tagFoundNotify();
+                }
+            }
+        }
+
+        public override void Flush()
+        {
+        }
+
+        public override void Dispose()
+        {
+            GC.SuppressFinalize(this);
+        }
+
+        private void Add(char c)
+        {
+            _startIndex++;
+
+            if (_startIndex == _buffer.Length - 1)
```

Comment:
might `_buffer.Length` be 0?

Reply:
It is derived from the `tag` string's length so it shouldn't be. I could add validation to `ctor` for check the tag is not empty